### DISCUSSION TITLE
CanOpenNode for ESP32

### DISCUSSION THREAD

**Discussion by thaanstad** [espressif/arduino-esp32#7953] [Ideas]
Thanks to the community for providing this awesome Arduino core for the ESP32 MCUs.  It has been very helpful on several of the projects I have worked on.  

I have ported an implementation of CANOpenNode for the ESP32 Wrover and created example code:
https://github.com/thaanstad/ESP32_CanOpenNode
https://github.com/CANopenNode/CANopenNode

Perhaps this would be useful for someone else developing a project utilizing the ESP32 and CAN communications.  I can do some work to implement this in the ESP32 Arduino core if that is desired.  

I did have some issues implementing the library in Arduino due to the #include requirements for Arduino.  For instance, the library utilizes `#include "301/CO_driver.h"` and I had to change these too `#include "../301/CO_driver.h"`.  These changes were throughout the library and this will make it difficult to keep up to date with future releases of the CanOpenNode library.  I'd greatly appreciate any suggestions regarding how to handle this, so it is easier to maintain in the future.

**Comment by mrengineer7777** [comment]
Nice!  Looks helpful.  Consider making this a library.
I believe you can add `component.mk` and/or `CMakeLists.txt` to your folders so that you can reference the header files by just the name.  See the library "ArduinoJson" for an example.

**Reply by mrengineer7777** [reply]
I would also encourage you to NOT add it to the Arduino core.  Getting changes added here is very slow.

**Comment by thaanstad** [comment]
@mrengineer7777 Thanks for your advice.  I'll look into the CMakeLists.txt.
